src/Http/Http/src/Features/ResponseBodyPipeFeature.cs

```diff
@@ -31,19 +35,31 @@ public PipeWriter Writer
                 }
 
                 if (_internalPipeWriter == null ||
-                    !object.ReferenceEquals(_internalPipeWriter.InnerStream, _context.Response.Body))
+                    !ReferenceEquals(_streamInstanceWhenWrapped, _context.Response.Body))
                 {
-                    _internalPipeWriter = new StreamPipeWriter(_context.Response.Body);
-                    _context.Response.RegisterForDispose(_internalPipeWriter);
+                    _streamInstanceWhenWrapped = _context.Response.Body;
+                    _internalPipeWriter = PipeWriter.Create(_context.Response.Body);
+
+                    _context.Response.OnCompleted((self) => ((ResponseBodyPipeFeature)self).Complete(), this);
                 }
 
                 return _internalPipeWriter;
             }
             set
             {
                 _userSetPipeWriter = value ?? throw new ArgumentNullException(nameof(value));
+                // REVIEW: Should we do this (vvvv) or just get rid of the setter?
                 // TODO set the response body Stream to an adapted pipe https://github.com/aspnet/AspNetCore/issues/3971
             }
         }
+
+        private Task Complete()
+        {
+            if(_internalPipeWriter != null && ReferenceEquals(_context.Response.Body, _streamInstanceWhenWrapped))
```

Comment:
Should we complete the active `PipeWriter` or every `PipeWriter` wrapper we ever created? Right now this code only completes a `PipeWriter` if the stream it wraps is still the active body stream.

Reply:
Ping @davidfowl @jkotalik

Reply:
I think we have to call complete on all wrappers created. Otherwise who else will clean it up?

Reply:
Yeah, that's my thought too.